### PR TITLE
Applied case sensitivity settings to KeywordsHelper

### PR DESCRIPTION
**Pull request by abbasc52**
Extended case sensitivity settings to KeywordHelper

### CODE REVIEW

**Comment by abbasc52** [comment]
@StefH, I am unable to find the right place to add test cases for this, could you guide me?

**Comment by StefH** [comment]
@abbasc52 
1. Changing this could break some existing usage. So in case this functionality is added, a new setting `KeywordsAreCaseSensitive` is needed.

2. Add a new test here `\Parser\KeywordsHelperTests`

**Comment by abbasc52** [comment]
@StefH Addressed your comment and added test cases, please review again

**Comment by abbasc52** [comment]
@StefH when can we merge/release this PR?

**Comment by StefH** [comment]
> @StefH when can we merge/release this PR?

I'll take a look.

**Comment by StefH** [comment]
because of PR https://github.com/zzzprojects/System.Linq.Dynamic.Core/pull/755, this PR needs to be updated....

**Comment by abbasc52** [comment]
@StefH I have updated the PR. Although looking at your PR, you seem to have fixed case issue with custom types. This PR will only help people make Context keywords case sensitive via config . I will leave it to you if you want to complete this PR.

**Comment by abbasc52** [comment]
My original goal of this PR was to make predefined and custom types case sensitive :)

File: src/System.Linq.Dynamic.Core/Parser/KeywordsHelper.cs [hunk 1]
```diff
@@ -64,6 +47,25 @@ public KeywordsHelper(ParsingConfig config)
     {
         _config = Check.NotNull(config);
 
+        _keywordMapping = new(config is { AreKeywordsCaseSensitive: true } ? StringComparer.Ordinal : StringComparer.OrdinalIgnoreCase)
```
Comment (StefH) [inline src/System.Linq.Dynamic.Core/Parser/KeywordsHelper.cs:50]: Just use `config.AreKeywordsCaseSensitive`

File: test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+using System.Linq.Dynamic.Core.Parser;
+using Xunit;
+
+namespace System.Linq.Dynamic.Core.Tests.Parser
+{
+    [DynamicLinqType]
+    public class TestClass
+    {
+        public string Hello { get; set; }
+    }
```
Comment (StefH) [inline test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs:14]: add newline

File: test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    public class KeywordsHelperTests
+    {
+
+        public KeywordsHelperTests()
+        {
+
```
Comment (StefH) [inline test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs:20]: no newline

File: test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+namespace System.Linq.Dynamic.Core.Tests.Parser
+{
+    [DynamicLinqType]
+    public class TestClass
+    {
+        public string Hello { get; set; }
+    }
+    public class KeywordsHelperTests
+    {
+
```
Comment (StefH) [inline test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs:17]: no newline

File: test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        [InlineData("testClass", false)]
+        [InlineData("nonExisting", false)]
+        public void TryGetValue_WithCaseSensitive_ReturnsResultAsExpected(string name, bool expected)
+        {
+            // Arrange
+            var keywordsHelper = this.CreateKeywordsHelper( new ParsingConfig { AreKeywordsCaseSensitive = true});
+            object type = null;
+
+            // Act
+            var result = keywordsHelper.TryGetValue(
```
Comment (StefH) [inline test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs:41]: use 1 line

File: test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        {
+          
+        }
+
+        private KeywordsHelper CreateKeywordsHelper(ParsingConfig config)
+        {
+            return new KeywordsHelper(config);
+        }
+
+        [Theory]
```
Comment (StefH) [inline test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs:28]: just create 1 unit test which tests all keywords + context-keywords and combine case-sensitive and case-insensitive:


``` c#
[Theory]
[InlineData("it", true, true)]
[InlineData("it", false, true)]
[InlineData("IT", true, false)]
[InlineData("IT", false, true)]
// and more
public void TryGetValue(string name, bool areKeywordsCaseSensitive, bool expected)
{
  // Arrange
  var config = new ParsingConfig
  {
    AreKeywordsCaseSensitive = areKeywordsCaseSensitive
  };
  var keywordsHelper = this.CreateKeywordsHelper(config);

  // Act
  var result = keywordsHelper.TryGetValue(name, out _);

  // Assert 
  result.Should().Be(expected);
}
```

**Review by StefH — CHANGES_REQUESTED** [review]

**Comment by StefH** [comment]
@abbasc52 
See my review comments.

**Comment by abbasc52** [comment]
@StefH fixed the review comments

File: test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        [InlineData("nonExisting", false, false)]
+        [InlineData("it", true, true)]
+        [InlineData("IT", true, false)]
+        [InlineData("TestClass", true, true)]
+        [InlineData("testClass", true, false)]
+        [InlineData("nonExisting", true ,false)]
+        public void TryGetValue_WithCaseSensitiveSettings_ReturnsResultAsExpected(string name, bool areKeywordsCaseSensitive,bool expected)
+        {
+            // Arrange
+            var keywordsHelper = this.CreateKeywordsHelper(new ParsingConfig { AreKeywordsCaseSensitive = areKeywordsCaseSensitive });
```
Comment (StefH) [inline test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs:41]: `this` i not needed

File: test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        [InlineData("IT", false, true)]
+        [InlineData("TestClass", false, true)]
+        [InlineData("testClass", false, false)]
+        [InlineData("nonExisting", false, false)]
+        [InlineData("it", true, true)]
+        [InlineData("IT", true, false)]
+        [InlineData("TestClass", true, true)]
+        [InlineData("testClass", true, false)]
+        [InlineData("nonExisting", true ,false)]
+        public void TryGetValue_WithCaseSensitiveSettings_ReturnsResultAsExpected(string name, bool areKeywordsCaseSensitive,bool expected)
```
Comment (StefH) [inline test/System.Linq.Dynamic.Core.Tests/Parser/KeywordsHelperTests.cs:38]: Do a format file to align the source code (adding some spaces)

**Review by StefH — CHANGES_REQUESTED** [review]

**Comment by StefH** [comment]
@abbasc52
See my review comments.

**Comment by StefH** [comment]
@abbasc52
See my review comments.

**Comment by StefH** [comment]
Hello @abbasc52, can you please take a look at my review comments?

**Comment by StefH** [comment]
Hello @abbasc52, can you please take a look at my review comments?

**Comment by StefH** [comment]
Hello @abbasc52, can you please take a look at my review comments?

**Pull request closed without merging**